Try: Allow vertical inserter in the nav block.

jasmussen opened the pull request:
Fixes #28313.

This PR refactors the margins of navigation menu items to have 8px space between them. This allows the vertical sibling inserter to work:

<img width="422" alt="Screenshot 2021-02-08 at 11 55 06" src="https://user-images.githubusercontent.com/1204802/107210685-c7d57280-6a04-11eb-87a6-3f76de0d1da9.png">

This is a bit of an opinionated change, because:

- It adds 8px margin between menu items. But counterpoint: zero margin between menu items is also opinionated, and themes can and should override this.
- The sibling inserter still doesn't work when a theme overrides the margin to be nil. Counterpoint: the vertical sibling inserter won't be a good experience when there's no clearance to invoke it in, so maybe it's good that isn't the case. And people can still insert and move blocks around.

Penny for your thoughts?

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

jasmussen committed:
Try: Allow vertical inserter in the nav block.

Fixes #28313.

aristath commented on packages/block-library/src/navigation-link/style.scss:
```diff
@@ -2,7 +2,7 @@
 	align-items: center;
 	position: relative;
-	margin: 0;
+	margin: 0 8px 0 0;
```
Could we do this in ems instead? 
```suggestion
	margin: 0 0.5em 0 0;
```

Since all other blocks use `em` values instead of `px` (with the exception of social links), I think it would make sense to keep it consistent

aristath commented on packages/block-library/src/navigation/theme.scss:
```diff
@@ -7,5 +7,5 @@
 // Overrides generic ".entry-content li" styles on the front end.
 .wp-block-navigation-link.wp-block-navigation-link {
-	margin: 0;
+	margin: 0 8px 0 0;
```
Same as above, it would probably be best to use `em` values.

```suggestion
	margin: 0 0.5em 0 0;
```

jasmussen commented:
Ems! Sure!

aristath reviewed (approved):
Thank you! LGTM ❤️

jasmussen commented:
I actually discovered a small issue with the CSS I removed, so I'm going to rebase, and fix that up, and then ask for a last sanity check. But thanks so much for the review so far Ari!

<img width="917" alt="Screenshot 2021-02-09 at 11 32 31" src="https://user-images.githubusercontent.com/1204802/107351175-9626df00-6aca-11eb-8509-781da0c8a568.png">

jasmussen commented:
Alright, should be good:

<img width="390" alt="Screenshot 2021-02-09 at 11 56 09" src="https://user-images.githubusercontent.com/1204802/107353900-dc317200-6acd-11eb-80e2-ac0ffe7dc4ee.png">

Note that those margins are spread across three files, and ideally they should be only in style.scss. I will follow up with a separate PR to merge those, so we can both simplify the markup, and have only a single file to override if this is to be customized.

aristath commented:
> Note that those margins are spread across three files, and ideally they should be only in style.scss. I will follow up with a separate PR to merge those, so we can both simplify the markup, and have only a single file to override if this is to be customized.

The styles in `theme.scss` are identical to the ones in `style.scss`... The only difference is that style.scss has `.wp-block-navigation-link` while in theme.scss the selector is duplicated as `.wp-block-navigation-link.wp-block-navigation-link` for increased specificity.
Perhaps we could remove the `theme.scss` styles completely? If needed we can increase the specificity in the `style.scss` rule because yeah, I agree that this duplication doesn't make a lot of sense 🤔

jasmussen commented:
That was my thought exactly.

jasmussen commented:
Created https://github.com/WordPress/gutenberg/pull/28878 as an immediate followup.

jasmussen commented:
I'm not sure why the tests keep failing. I'm getting this:

```
There are local uncommitted changes after one or both of 'npm install' or 'npm run docs:build'!
diff --git a/package-lock.json b/package-lock.json
index 500fa33..fb1291b 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -45755,9 +45755,9 @@
 					"dev": true
 				},
 				"bl": {
-					"version": "4.0.4",
-					"resolved": "https://registry.npmjs.org/bl/-/bl-4.0.4.tgz",
-					"integrity": "sha512-7tdr4EpSd7jJ6tuQ21vu2ke8w7pNEstzj1O8wwq6sNNzO3UDi5MA8Gny/gquCj7r2C6fHudg8tKRGyjRgmvNxQ==",
+					"version": "4.1.0",
+					"resolved": "https://registry.npmjs.org/bl/-/bl-4.1.0.tgz",
+					"integrity": "sha512-1W07cM9gS6DcLperZfFSj+bWLtaPGSOHWhPiGzXmvVJbRLdG82sH/Kn8EtW1VqWVA54AKf2h5k5BbnIbwF3h6w==",
 					"dev": true,
 					"requires": {
 						"buffer": "^5.5.0",
```

not sure why this PR would cause that.

youknowriad commented:
@jasmussen just commit them here (or separately to master) as you wish :)

jasmussen commented:
The margin changes are potentially a bit sensitive, so I figured I'd keep things separate ☺️

ItsJonQ commented:
> I'm not sure why the tests keep failing. I'm getting this:

@jasmussen I don't think it's this PR 🙏 . It looks like the recent commits into master have been failing for this reason:

https://github.com/WordPress/gutenberg/commits/master

shaunandrews commented:
This makes sense for the horizontal variation, but for the vertical variation (which is also used on the new Navigation screen) this PR adds some unneeded margin:

<img width="648" alt="image" src="https://user-images.githubusercontent.com/191598/107536312-a228a480-6b8f-11eb-978f-778394220732.png">

<img width="454" alt="image" src="https://user-images.githubusercontent.com/191598/107536337-a94fb280-6b8f-11eb-929d-b36c220cc8e5.png">

jasmussen commented:
Great catch, I'll address that. Thank you Shaun.

jasmussen commented:
Alright, in https://github.com/WordPress/gutenberg/pull/28833/commits/2eae54dfc272810afd7cdc3d478411ec316ea5cc I pushed some tweaks to margins.

But I also discovered that the rabbit hole goes a bit deeper now that the page list has landed. It doesn't make sense to introduce the margin only for the separate menu items. So I'm pausing this one for a bit to give it a bit more time in the oven. 

In #28265 @tellthemachines wrote:

> Speaking of which, this block rendering a ul container is inconsistent with the current Nav block markup which expects all its children to be li elements. I'm expecting that to be fixed by #24644 / #24364. The Pages block shouldn't be contained in a li as it can be used separately from the Nav block.

It sounds like that might simplify some of the CSS classes, so I have to write less edgecases (adding complexity). Once that's further along, it's probably a good time to rebase this one and pick it up again.

jasmussen commented:
Waiting for #29465.

jasmussen commented:
Alright, this one has been unblocked by the recent refactor of the navigation block. I've rebased, squashed, and simplified, so the changeset is smaller now. The testing instructions are the same — you should see the vertical inserter between menu items in the editor. And you should see a .5em margin between items front and backend:

<img width="815" alt="Screenshot 2021-03-11 at 09 03 14" src="https://user-images.githubusercontent.com/1204802/110755228-01d6a600-8249-11eb-89e1-a5594a72a0eb.png">

<img width="672" alt="Screenshot 2021-03-11 at 09 03 20" src="https://user-images.githubusercontent.com/1204802/110755229-0307d300-8249-11eb-9478-6d53215cf866.png">

<img width="551" alt="Screenshot 2021-03-11 at 09 03 26" src="https://user-images.githubusercontent.com/1204802/110755236-04d19680-8249-11eb-832a-c04a2528cfd9.png">

jasmussen commented:
But because it changed a bit, I would appreciate a re-review or sanity check!

draganescu reviewed (approved):
I gave this a spin and it feels very good. It has no visible effect in the navigation editor or the vertical variation of the block.
